### ISSUE TITLE
RedisSessionStore - hkeys

### ISSUE THREAD

**Issue by lvnar** [rejectedsoftware/vibe.d#1412]
I cannot use de hkeys function on redis. There is an exception when I run it:

> core.exception.AssertError@../.dub/packages/vibe-d-0.7.27/vibe-d/source/vibe/db/redis/redis.d(1150): Assertion failure
Program exited with code 1

Any idea?

**Comment by etcimon** [comment]
That's because you used a `RedisReply!T` and didn't consume it completely in a previous connection. Make sure you do `popFront()` on each field until `.front` is `null`

**Comment by lvnar** [comment]
I'm using popFront(), while(x.hasNext), but the error is still there when I try to use RedisReply again.

**Comment by etcimon** [comment]
Do you have a code sample?

**Comment by lvnar** [comment]
`auto scan = redisDB.keys(pattern);

while(scan.hasNext) {
	keys ~= scan.front;
	scan.popFront();
}
foreach (string key; keys) {
	try {
		keyValue = [key, redisDB.get(key)];
		result ~= keyValue;
	}
	catch (Exception e){
		auto res = redisDB.hgetAll(key);
		while(res.hasNext){
			res.popFront();
		}
	}
}`

There, on hgetAll is where the exception appears:
> core.exception.AssertError@../.dub/packages/vibe-d-0.7.27/vibe-d/source/vibe/db/redis/redis.d(1150): Assertion failure

**Comment by etcimon** [comment]
I think hasNext will return false while there is one item left to pop, so you're leaving one item hanging. You should use `foreach(key; scan)` instead it's less error prone

**Comment by etcimon** [comment]
The problem is that your first `RedisReply` from the `keys` call is still using the connection while you try to acquire the connection in a second `RedisReply` from the `hgetAll` call. You can simply enclose the first part in a scope like this: 

```D
{
   auto scan = redisDB.keys(pattern);

   while(scan.hasNext) {
      keys ~= scan.front;
      scan.popFront();
   }
}
```

**Comment by etcimon** [comment]
You should avoid making calls that can throw inside `catch` clauses, and make sure you don't use try/catch as a replacement for if/else because it takes a while to populate a stack trace.

**Comment by s-ludwig** [comment]
I've opened a fix (#1416), which is a slightly more refined version of the fix in #1396. It doesn't have a test case, yet. I'll add that before pulling.

**Comment by lvnar** [comment]
Thaks for the advices, were useful.
